src/Silicon.cpp

```diff
@@ -0,0 +1,506 @@
+            _distortions[i * _ny + j][n].y = y;
+        }
+#ifdef DEBUGLOGGING
+        //Test print out of distortion for central pixel
+        int i = 4;
+        int j = 4;
+        for (int n=0; n < _nv; n++) {
+            xdbg<<"n = "<<n<<", x = "<<_distortions[i * _ny + j][n].x * numElec
+                <<", y = "<<_distortions[i * _ny + j][n].y * numElec<<std::endl;
+        }
+#endif
+    }
+
+#if 0
+    // Got a start on implementing the absorption length lookup,
+    // but couldn't get the python - C++ wrapper working, so I went
+    // back to the analytic function - Craig Lage 17-Feb-17
+    double Silicon::AbsLength(double lambda)
+    {
+        // Looks up the absorption length and returns
+        // an interpolated value
+        int nminus;
+        double aminus, aplus, dlambda;
+        dlambda = _abs_data[2] - _abs_data[0];
+        // index of
+        nminus = (int) ((lambda - _abs_data[0]) / dlambda) * 2;
+        if (nminus < 0) return _abs_data[1];
+        else if (nminus > _nabs - 4) return _abs_data[_nabs - 1];
+        else return _abs_data[nminus+1] +
+            (lambda - _abs_data[nminus]) * (_abs_data[nminus+3] - _abs_data[nminus+1]);
+    }
+#endif
+
+    template <typename T>
+    void Silicon::updatePixelDistortions(ImageView<T> target)
+    {
+        // This updates the pixel distortions in the _imagepolys
+        // pixel list based on the amount of charge in each pixel
+        // This distortion assumes the electron is created at the
+        // top of the silicon.  It mus be scaled based on the conversion depth
+        // This is handled in insidePixel.
+
+        int nxCenter = (_nx - 1) / 2;
+        int nyCenter = (_ny - 1) / 2;
+
+        // Now add in the displacements
+        int minx = target.getXMin();
+        int miny = target.getYMin();
+        int maxx = target.getXMax();
+        int maxy = target.getYMax();
+
+        // Now we cycle through the _imagepolys array
+        // and update the pixel shapes
+        for (int ix=minx; ix<maxx; ix++) {
+            for (int iy=miny; iy<maxy; iy++) {
+                int index = (ix - minx) * (maxy - miny + 1) + (iy - miny);
+                // First set the _imagepoly polygon to an undistorted polygon
+                _imagepolys[index] = _emptypoly;
+                // Now add in the distortions
+                for (int i=nxCenter-_qDist; i<nxCenter+_qDist+1; i++) {
+                    for (int j=nyCenter-_qDist; j<nyCenter+_qDist+1; j++) {
+                        int chargei = ix + i - nxCenter;
+                        int chargej = iy + j - nyCenter;
+                        if ((chargei < minx) || (chargei > maxx) ||
+                            (chargej < miny) || (chargej > maxy))
+                            continue;
+
+                        double charge = target(chargei,chargej);
+                        if (charge < 100.0) // Don't bother if less than 100 electrons
```

Comment:
Note, as part of the changes I just made, I switched this to just `if (charge == 0.0) continue`.  At least for the runs we are doing now, the difference between 100 and 0 wasn't significant, so I just removed the approximation entirely.